### ISSUE TITLE
#[warn(dead_code)] showing unnecessary x is never used warnings

### ISSUE THREAD

**Issue by huss4in** [rust-lang/rustlings#2104]
Since testing in some exercises is done using tests modules, annoying `function 'x' is never used warning` shows while and after solving the exercise...
![image](https://github.com/user-attachments/assets/c0d91e80-f251-4174-a299-f19002de02e4)

Which could sometimes be annoying or even come before/hide more relevant code lints related to the main exercise issue:

Before: 
![before](https://github.com/user-attachments/assets/d40e4418-e6ff-4543-9ce8-c6859e4f9002)
After: 
![after](https://github.com/user-attachments/assets/8de1a810-dadf-4f39-b1aa-a39fba419076)

A simple way to make it look cleaner is by adding `#![allow(dead_code)]` to all files which have `mod tests {`

```rust
| #![allow(dead_code)]

  fn fill_vec(vec: Vec<i32>) -> Vec<i32> {
      let mut vec = vec;

      vec.push(88);

      vec
  }
  ...
```

Or in a more generic/centralized way in `Cargo.toml`

```toml
[lints.rust]
usued = "allow"
```
```[tasklist]
### Solutions
- [ ] https://github.com/rust-lang/rustlings/pull/2105
- [ ] https://github.com/rust-lang/rustlings/pull/2106
```

**Comment by mo8it** [comment]
Allowing dead code is just a workaround. The actual solution would be to tell Rust-Analyzer to consider the usage in tests.

Rustlings itself does so by passing `--profile test` to the used Cargo commands. I will see if I can pass this option to Rust-Analyzer.

**Comment by mo8it** [comment]
@alibektas is the `rust-analyzer.toml` ready for usage? This issue can be resolved by setting `check.extraArgs` to `["--profile", "test"]`.

**Comment by alibektas** [comment]
To answer your question : It is honestly not completely ready as we keep changing config classes a lot but I think in your case it is safe to say that we are not going to make any changes to config keys `check.*`.

**Comment by huss4in** [comment]
Thanks for looking into the issue!

I would just like to shed some lights on the fact that using `--profile test` doesn't solve all cases
Example in `enums2`:
![image](https://github.com/user-attachments/assets/cfd02e17-0c1d-4c93-a9e1-b9c962d6ad16)

Compared to `allow(dead_code)`:
![image](https://github.com/user-attachments/assets/bb9872aa-2cfa-4968-9f5f-7df9378798c3)

I understand that you might not want to globally turn off `dead_code` lints.. or `unused` which includes `dead_code` and more declared but not used things..
![image](https://github.com/user-attachments/assets/aa951654-f03d-4aec-8d7d-06d8a100f4fd)
As it might possibly be helpful in some exercises..

However, I just believe that they might not be that useful/important for someone new and trying to learn and might cause some confusion.
Let me know what you think?

And how do you suggest a better solution?
Thanks!

**Comment by mo8it** [comment]
@huss4in Thanks for elaborating. You are right. I tried to deal with these cases by adding individual allows. But it is easy to miss one. So I added the lint in https://github.com/rust-lang/rustlings/commit/0513660b05e8dd45ba7bb25fff89b4fd089b14ea (and corrected its position in https://github.com/rust-lang/rustlings/commit/4ffce1c2971d127149e5530a94504664d9d5bc50).

I am against allowing the whole `unused` group though. It has very useful lints.
